tools/ExampleExtractor/ExtractorAndTesterUsersGuide.md

```diff
@@ -0,0 +1,483 @@
+    ;
+annotation_directive
+    : template
+    | name
+    | replace_ellipsis
+    | expected_errors
+    | expected_warnings
+    | ignored_warnings
+    | expected_output
+    | expected_exception
+    ;
+```
+
+While an *example_annotation* must precede its corresponding example, it need not be immediately prior. The two can be separated by text and/or HTML comments. However, if multiple *example_annotation*s precede the same code block, the second and subsequent *example_annotation*s are ignored.
+
+Arbitrary horizontal whitespace is permitted between any two adjacent tokens.
+
+If multiple *annotation_directive*s contain the same name, the right-most directive prevails.
+
+The ordering of annotation directives is not significant.
+
+The C# spec contains two kinds of source-code examples: those within an \*Example\* and those not, as follows:
+
+````
+> *Example*:
+>
+> <!-- Example: { … } -->
+> ```csharp
+> // source code goes here
+> ```
+>
+> *end example*
+````
+
+In this case, the *example_annotation* is preceded by `> ` to match the example’s formatting. However, in the following case, it is not, as no \*Example\* is present:
+
+````
+<!-- Example: { … } -->
+```csharp
+// source code goes here
+```
+````
+
+### Templates
+
+An annotation’s *template* is used to select certain compiler options, and to provide supporting machinery, as needed. This *annotation_directive* is required.
+
+```ANTLR
+template
+    : 'template' ':' template_name
+    ;
+template_name
+    : '"standalone-console"'  // actually, a JSON_string_value with this content
+    | '"standalone-lib"'      // actually, a JSON_string_value with this content
+    | '"code-in-main"'        // actually, a JSON_string_value with this content
+    ;
+```
+
+The template `standalone-console` indicates that the example is an application. For example:
+
+````
+> <!-- Example: {template:"standalone-console", …} -->
+> ```csharp
+> class Hello
+> {
+>     static void Main()
+>     {
+>         …
+>     }
+> }
+> ```
+````
+
+The template `standalone-lib` indicates that the example is a class library. For example:
+
+````
+> <!-- Example: {template:"standalone-lib", … } -->
+> ```csharp
+> class Class1
+> {
+>     static void Test(bool status)
+>     {
+>         …
+>     }
+> }
+> ```
+````
+
+The template `code-in-main` indicates that the example needs to be wrapped inside an entry-point method inside a class. For example, the example:
+
+````
+> <!-- Example: {template:"code-in-main", … } -->
+> ```csharp
+> int[][] pascals = 
+> {
+>     new int[] {1},
+>     new int[] {1, 1},
+>     new int[] {1, 2, 1},
+>     new int[] {1, 3, 3, 1}
+> };
+> ```
+````
+
+gets transformed into the following application:
+
+````
+class Program
+{
+    static void Main()
+    {
+        int[][] pascals = 
+{
+    new int[] {1},
+    new int[] {1, 1},
+    new int[] {1, 2, 1},
+    new int[] {1, 3, 3, 1}
+};
+    }
+}
+````
+
+### Example Names
```

Comment:
It would be good to recommend that example names are valid C# identifiers. (I've been renaming some as I go along.)

At the moment it's not strictly required, but I can see it being useful later, potentially (e.g. for namespaces).